# Introduction to Transfer Learning

## Goal

The purpose of this notebook is to demonstrate the concept of transfer learning in computer vision.

Instead of training a neural network from scratch, we reuse a pretrained model and adapt it to a new task.

In [1]:
import torch
import torch.nn as nn
from torchvision import models

## Loading a Pretrained ResNet18 Model

ResNet18 is a convolutional neural network pretrained on the ImageNet dataset.

We can reuse its learned features instead of training a completely new model.

In [2]:
model = models.resnet18(weights="DEFAULT")

print(type(model))

Downloading: "https://download.pytorch.org/models/resnet18-f37072fd.pth" to C:\Users\teter/.cache\torch\hub\checkpoints\resnet18-f37072fd.pth


100.0%


<class 'torchvision.models.resnet.ResNet'>


## Exploring the Model

Let's inspect the architecture and number of parameters in the pretrained ResNet18 model.

In [4]:
total_params = sum(p.numel() for p in model.parameters())

print(f"Total parameters: {total_params:,}")

Total parameters: 11,689,512


In [5]:
print(model.fc)

Linear(in_features=512, out_features=1000, bias=True)


## Freezing the Feature Extractor

The pretrained layers already contain useful visual knowledge.

Instead of training the entire network again, we freeze the pretrained layers and train only the final classification layer.

In [6]:
for param in model.parameters():
    param.requires_grad = False

In [7]:
trainable_before = sum(
    p.numel()
    for p in model.parameters()
    if p.requires_grad
)

print(f"Trainable parameters before replacing the classifier: {trainable_before:,}")

Trainable parameters before replacing the classifier: 0


## Replacing the Classification Head

The original ResNet18 model was trained to classify 1000 ImageNet categories.

For a new task, we replace the final classification layer with a new one that matches the number of target classes.

In [8]:
model.fc = nn.Linear(
    model.fc.in_features,
    2
)

print(model.fc)

Linear(in_features=512, out_features=2, bias=True)


In [9]:
trainable_after = sum(
    p.numel()
    for p in model.parameters()
    if p.requires_grad
)

print(f"Trainable parameters after replacing the classifier: {trainable_after:,}")

Trainable parameters after replacing the classifier: 1,026


In [10]:
print(f"Total parameters: {total_params:,}")
print(f"Trainable parameters: {trainable_after:,}")
print(f"Percentage trainable: {100 * trainable_after / total_params:.4f}%")

Total parameters: 11,689,512
Trainable parameters: 1,026
Percentage trainable: 0.0088%


## Results

The pretrained ResNet18 model contains 11,689,512 parameters.

After freezing the feature extractor and replacing the classification head, only 1,026 parameters remain trainable.

This represents only 0.0088% of the total model parameters.

Therefore, transfer learning allows us to adapt a large neural network to a new task while training only a very small number of parameters.

## Real-World Applications

Transfer learning is widely used in computer vision and biomedical imaging.

For example, a model pretrained on ImageNet can be adapted to classify medical images or rat brain images using only a small amount of labeled data.

This significantly reduces training time and data requirements.

# Conclusion

In this notebook, we explored the concept of transfer learning using a pretrained ResNet18 model.

We demonstrated:
- loading a pretrained model;
- freezing pretrained layers;
- replacing the classification head;
- reducing the number of trainable parameters.

Transfer learning enables efficient adaptation of deep neural networks to new computer vision tasks.